ownCloudAppShared/Intent/CreateFolderIntentHandler.swift

```diff
@@ -0,0 +1,106 @@
+//
+//  CreateFolderIntentHandler.swift
+//  ownCloudAppShared
+//
+//  Created by Matthias Hühne on 31.07.19.
+//  Copyright © 2019 ownCloud GmbH. All rights reserved.
+//
+
+/*
+ * Copyright (C) 2019, ownCloud GmbH.
+ *
+ * This code is covered by the GNU Public License Version 3.
+ *
+ * For distribution utilizing Apple mechanisms please see https://owncloud.org/contribute/iOS-license-exception/
+ * You should have received a copy of this license along with this program. If not, see <http://www.gnu.org/licenses/gpl-3.0.en.html>.
+ *
+ */
+
+import UIKit
+import Intents
+import ownCloudSDK
+
+@available(iOS 13.0, watchOS 6.0, *)
+public class CreateFolderIntentHandler: NSObject, CreateFolderIntentHandling {
+
+	var itemTracking : OCCoreItemTracking?
+
+	public func handle(intent: CreateFolderIntent, completion: @escaping (CreateFolderIntentResponse) -> Void) {
+		if AppLockHelper().isPassCodeEnabled {
+			completion(CreateFolderIntentResponse(code: .authenticationRequired, userActivity: nil))
+		} else {
+			if let path = intent.path, let uuid = intent.account?.uuid, let name = intent.name {
+				let accountBookmark = OCBookmarkManager.shared.bookmark(for: uuid)
+
+				if let bookmark = accountBookmark {
+					OCCoreManager.shared.requestCore(for: bookmark, setup: nil, completionHandler: { (core, error) in
+						if error == nil, let core = core {
+							self.itemTracking = core.trackItem(atPath: path, trackingHandler: { (error, item, isInitial) in
+								if let targetItem = item {
+									if core.createFolder(name, inside: targetItem, options: nil, resultHandler: { (error, _, item, _) in
+										if error != nil {
```

Comment:
@hosy I think, these nested conditions can be simplified and number of completion handler calls can be reduced for the sake of better readability

Reply:
Looks strange, but there are different return codes for the completion handlers. I think cannot be reduced, if I want to give the correct failure feedback top the user.

Reply:
@hosy I think it could be at least flatten a bit, e.g. like this:
```suggestion
						if error == nil, let core = core {
							self.itemTracking = core.trackItem(atPath: path, trackingHandler: { (error, item, isInitial) in

								guard let targetItem = item else {
									completion(CreateFolderIntentResponse(code: .pathFailure, userActivity: nil))
									return
								}

								let progress = core.createFolder(name, inside: targetItem, options: nil, resultHandler: { (error, _, item, _) in
									if error != nil {
										completion(CreateFolderIntentResponse(code: .failure, userActivity: nil))
									} else {
										completion(CreateFolderIntentResponse.success(path: targetItem.path ?? ""))
									}
								})

								if progress == nil {
									completion(CreateFolderIntentResponse(code: .failure, userActivity: nil))
								}

								if isInitial {
									self.itemTracking = nil
								}
							})

						} else {
							completion(CreateFolderIntentResponse(code: .failure, userActivity: nil))
						}
```